javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/RemappingUrlConnection.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling;
+
+import static io.opentelemetry.javaagent.tooling.ShadingRemapper.rule;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.security.Permission;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import net.bytebuddy.jar.asm.ClassReader;
+import net.bytebuddy.jar.asm.ClassWriter;
+import net.bytebuddy.jar.asm.commons.ClassRemapper;
+
+public class RemappingUrlConnection extends URLConnection {
+  // We need to prefix the names to prevent the gradle shadowJar relocation rules from touching
+  // them. It's possible to do this by excluding this class from shading, but it may cause issue
+  // with transitive dependencies down the line.
+  private static final ShadingRemapper remapper =
+      new ShadingRemapper(
+          rule("#io.opentelemetry.api", "#io.opentelemetry.javaagent.shaded.io.opentelemetry.api"),
+          rule(
+              "#io.opentelemetry.context",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.context"),
+          rule(
+              "#io.opentelemetry.extension.aws",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.extension.aws"),
+          rule("#java.util.logging.Logger", "#io.opentelemetry.javaagent.bootstrap.PatchLogger"),
+          rule("#org.slf4j", "#io.opentelemetry.javaagent.slf4j"));
+
+  private final JarFile delegateJarFile;
+  private final JarEntry entry;
+
+  private InputStream cachedResult;
```

Comment:
Is only one `URLConnection` ever instantiated? Not sure when `openConnection` is called during classload

Reply:
Certainly not one, why do you ask?

Reply:
If not once, I guess this cache is never used? Should it be a static cache?

Reply:
A, that's the catch :) it IS read/called at least 2 times by `URLClassoader`. Old `AgentClassLoader` also had an optimisation for this double reading.

Reply:
Caching `Inputstream` this way is weird. As streams can be read only once without using mark/reset then this caching would make sense only when someone opens the stream and doesn't use it at all and then reopens the stream and actually reads it. Perhaps this isn't really needed any more? If it is needed then caching byte array and returning a new `ByteArrayInputStream` from `getInputStream` would be nicer.

Reply:
`handler.getInputStream()` is called twice:

1. In `jdk.internal.loader.URLClassPath.Loader#getResource(java.lang.String, boolean)`, where exactly "someone opens the stream and doesn't use it at all" happens
2. in `jdk.internal.loader.Resource#cachedInputStream`, when stream is actually used afterwards.

But I agree with you that caching byte arrays makes more sense, will do it.